### PR TITLE
Config restructuring.

### PR DESCRIPTION
**Pull request by NegativeMjark**
This PR is intended to drastically reduce the number of options that are configured through both commandline arguments and through yaml. The yaml config is promoted to being the primary config file format and the automatic config generator will now include comments to document the yaml config options. The "--generate-config" option will no longer be able to incrementally add new defaults to an exiting config as there isn't an easy way to round trip a yaml file while preserving comments, therefore --generate-config will error if the file already exists.

This PR includes a number of other improvements:
* auto-generated key_ids have a random suffix
* A yaml logging config is automatically generated.
* Both enable_registration and disable_registration are supported.
* Durations can be specified in hours, days and weeks as well as in milliseconds.
* The default bind port will be automatically derived from the server_name.

### CODE REVIEW

**Comment by NegativeMjark** [comment]
Also the database config and perspective config are included within the config yaml itself rather than being separate files.

File: README.rst [hunk 1]
```diff
@@ -170,7 +170,7 @@ traditionally used for convenience and simplicity.
 
 The advantages of Postgres include:
 
- * significant performance improvements due to the superior threading and
+ * significant performannce improvements due to the superior threading and
```
Comment (erikjohnston) [inline README.rst]: `nnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn`

File: synapse/config/metrics.py [hunk 2]
```diff
@@ -17,20 +17,17 @@
...
-        )
+    def default_config(self, config_dir_path, server_name):
+        return """\
+        ## Metrics ###
+
+        # Enable collection and rendering of performance metrics
+        enable_metrics: False
+
+        # Separate port to accept metrics requests on (on localhost)
+        metrics_port: ~
```
Comment (erikjohnston) [inline synapse/config/metrics.py]: Maybe use `# matrix_port: 8081` instead?

File: synapse/config/key.py [hunk 3]
```diff
@@ -120,28 +125,8 @@ def generate_config(cls, args, config_dir_path):
                     "auto",
```
Comment (erikjohnston) [inline synapse/config/key.py]: Should this be something other than `auto`?

**Comment by erikjohnston** [comment]
Does it support the old file format correctly?

**Comment by NegativeMjark** [comment]
I've tested it against the old file format at bit.

**Comment by erikjohnston** [comment]
LGTM

**Comment by ara4n** [comment]
why do we want to stop --generate-config from being able to automatically add new config options? having the ability to autoupgrade folks' configs seemed like quite a nice way to avoid sysadmins having to manually port in new config changes, which can get tedious fast...

**Comment by NegativeMjark** [comment]
It's hard to add new keys or modify existing keys without losing any comments within the yaml. My feeling is that as we include more options within the config the comments added by us or by the admin will be more valuable than being able to run generate config to update an existing file.

**Comment by NegativeMjark** [comment]
Also now that we expose the yaml to the config code it is much easier to add or change keys in the config while preserving backwards compatibility because we don't have to include every config key we understand in the generated default config file.

File: README.rst (hunk 1 above)
Comment (ara4n) [inline README.rst]: if this causes pain, please suggest something better... surely postgres' multithreading should be helping us. Or are you objecting to the suggestion that performance overall has actually increased?

Comment (erikjohnston) [inline README.rst]: This change was an accident :)

Comment (ara4n) [inline README.rst]: oh, i'm blind. i thought you were growling...

**Comment by NegativeMjark** [comment]
I've gone back to allowing synapse to --generate-config with an existing config-file. It will auto generate missing key and logging files, but won't fill out any of the default keys.

**Pull request merged**